File: sorts/quickSort.py
# QuickSort in Python 2.7
# By Marshall Ehlinger
# For 2015 Analysis of Algorithms class
# Written as a protype for C code


def quickSort(array, lo, hi):
	if lo < hi:
		pivot = partition(array, lo, hi)
		quickSort(array, lo, pivot-1)
		quickSort(array, pivot+1, hi)

	

def partition(array, lo, hi):
	pivot = array[hi]
	for i in range(lo, hi):
		if array[i] <= pivot:
			swap(array, i, lo)
			lo += 1
	swap(array, lo, hi)
	return lo # return the 'centerpoint'/wall

def swap(a, first, second):
	temp = a[first]
	a[first] = a[second]
	a[second] = temp


########
# MAIN #
########
def main():
	with open("sorts.txt", "r") as data:
		array = data.readlines()
	array = array[0].split(",")
	for i in range (0, len(array)):
		array[i] = int(array[i])
	quickSort(array, 0, len(array)-1)
	print array

########
# TEST #
########	
main()

File: sorts/quickSort.c
/*
* QuickSort in C
* by Marshall Ehlinger
* for Algorithms Spring 2015
* Main method sorts a test array of integers
*/

#include <stdio.h>
#include <stdlib.h>

void swap(int *a, int firstIndex, int secondIndex) {
	int temp;
	temp = a[firstIndex];
	a[firstIndex] = a[secondIndex];
	a[secondIndex] = temp;
}

int partition(int *a, int lo, int hi) {
	int pivot = a[hi];
	int i;

	for (i = lo; i < hi; i++) {
		if (a[i] <= pivot) {
			swap(a, i, lo);
			lo++;
		}
	}
	swap(a, lo, hi);
	return lo;
}

void quickSort(int *a, int lo, int hi) {
	int pivot;
	if (lo < hi) {
		pivot = partition(a, lo, hi);
		quickSort(a, lo, pivot-1);
		quickSort(a, pivot+1, hi);
	}
}

void printArray(int a[], int arrayLen) {
	int i;
	for (i = 0; i < arrayLen; i++) {
		printf("%d ", a[i]);
	}
	printf("\n");
}

/////////////////
//     MAIN    //
/////////////////
int main() {
	int array[] = {2, 3, 5, 22, 1, 15, 4};
	int arrayLen = sizeof(array) / sizeof(array[0]);

	printArray(array, arrayLen);
	
	quickSort(array, 0, arrayLen-1);

	printArray(array, arrayLen);

}